src/Data/RadixNTree/Word8/Lazy.hs

```diff
@@ -0,0 +1,5076 @@
+          case f Nothing of
+            Nothing -> t
+            Just a  -> singleton_ step w s a
+
+
+
+{-# INLINE shape0 #-}
+shape0 :: (RadixTree a -> RadixTree a) -> Feed -> RadixTree a -> RadixTree a
+shape0 f (Feed feed) = \t0@(RadixTree mx t) ->
+  feed $ \step s ->
+    case step s of
+      More w z -> RadixTree mx $ shape_ f step w z t
+      Done     -> f t0
+
+{-# INLINE shape1 #-}
+shape1 :: (RadixTree a -> RadixTree a) -> Feed1 -> Radix1Tree a -> Radix1Tree a
+shape1 f (Feed1 w feed) =
+  feed $ \step -> shape_ f step w
+
+{-# INLINE shape_ #-}
+shape_
+  :: (RadixTree a -> RadixTree a)
+  -> (x -> Step Word8 x) -> Word8 -> x -> Radix1Tree a -> Radix1Tree a
+shape_ f step = go
+  where
+    go w s t =
+      case t of
+        Bin p l r
+          | beyond p w -> let !(RadixTree my dy) = f (RadixTree Nothing Nil)
+                          in case retip (fromStep step w s) my dy of
+                               Nil -> t
+                               dz  -> join
+                                        w dz
+                                        p t
+
+          | w < p      -> rebinL p (go w s l) r
+          | otherwise  -> rebinR p l (go w s r)
+
+        Tip arr mx dx -> goarr w s 0
+          where
+            goarr v z n
+              | v == indexByteArray arr n =
+                  let n' = n + 1
+                  in if n' >= sizeofByteArray arr
+                       then case step z of
+                              More u z' -> retip arr mx (go u z' dx)
+                              Done      -> let !(RadixTree my dy) = f (RadixTree mx dx)
+                                           in retip arr my dy
+
+                       else case step z of
+                              More u z' -> goarr u z' n'
+                              Done      ->
+                                let !(# !brr, !crr #) = splitByteArray 0 n' arr
+
+                                    !(RadixTree my dy) = f (RadixTree Nothing (Tip crr mx dx))
+
+                                in retip brr my dy
+
+              | otherwise =
+                  let !(RadixTree my dy) = f (RadixTree Nothing Nil)
+                  in case retip (fromStep step v z) my dy of
+                       Nil -> t
+                       dz  ->
+                         if n == 0
+                           then join
+                                  (indexByteArray arr 0) (Tip arr mx dx)
+                                  v dz
+
+                           else let !(# !brr, !crr #) = splitByteArray 0 n arr
+                                in Tip brr Nothing $
+                                     join
+                                       (indexByteArray crr 0) (Tip crr mx dx)
+                                       v dz
+
+        Nil       ->
+          let !(RadixTree my dy) = f (RadixTree Nothing Nil)
+          in retip (fromStep step w s) my dy
+
+
+
+{-# INLINE splitL0 #-}
+splitL0 :: Openness -> Feed -> RadixTree a -> (RadixTree a, RadixTree a)
```

Comment:
For strict radix trees the signature is
```haskell
splitL0 :: Openness -> Feed -> RadixTree a -> Split a a
```

Would it make sense to copy what `containers` do here and make strict version of `splitL0` also return lazy tuple so that users can switch with no cost? Strict version will be a wrapper around working returning strict tuple and just repackage it into a lazy tuple.

Reply:
The introduction of `Split` and `SplitLookup` was a very late change from my side, the thing that irked me was that the strict splits went through the trouble of creating two completely strict values and then at the very end threw in two boxes in the form of a lazy 2-tuple. I as such prefer a strict tuple because I expect all results in the strict module to be strict.

> copy what `containers` do ... so that users can switch with no cost

`Lazy` and `Strict` in this library do not mean the same thing as they do in `containers`. `containers`' definition is "everything in `Lazy` does not evaluate the value before putting into the tree, everything in `Strict` does", but their underlying implementation is spine-strict in both cases.

The difference between a split on a spine-lazy tree and a split on a spine-strict one is massive, so I don't see the need to have the two be directly interchangeable.